javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,164 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import java.io.Serializable;
+
+import static javaslang.collection.Arrays.*;
+
+/**
+ * A `bit-mapped trie` is a very wide and shallow tree (for integer indices the depth will be ≤ 6).
+ * Each node has a maximum of `32` children (configurable).
+ * Access to a given position is done by converting the index to a base 32 number and using each digit to descend down the tree.
+ * Modifying the tree is done similarly, but along the way the path is copied, returning a new root every time.
+ * Append inserts in the last leaf, or if the tree is full from the right, it adds another layer on top of it (the old root will be the first of the new one).
+ * Prepend is done similarly, but an offset is needed, because adding a new top node (where the current root would be the last node of the new root)
+ * shifts the indices by half of the current tree's full size. The `offset` shifts them back to the correct index.
+ *
+ * @author Pap Lőrinc
+ * @since 2.1.0
+ */
+final class BitMappedTrie<T> implements Serializable {
+    static int BRANCHING_BASE = 5;
+    static int branchingFactor()                   { return 1 << BRANCHING_BASE; }
+    static int firstDigit(int num, int depthShift) { return num >> depthShift; }
+    static int digit(int num, int depthShift)      { return lastDigit(firstDigit(num, depthShift)); }
+    static int lastDigit(int num)                  { return num & (-1 >>> -BRANCHING_BASE); }
+
+    private static final long serialVersionUID = 1L;
+
+    private static final BitMappedTrie<?> EMPTY = new BitMappedTrie<>(Arrays.empty(), 0, 0, 0);
```

Comment:
Why is `Arrays.empty()` a method and not a constant?

Why do we not initialize `EMPTY` with `new BitMappedTrie<>(new Object[0], 0, 0, 0);` and get the empty array like so when needed within BitMappedTrie: `EMPTY.array`?